### ISSUE TITLE
Add data classes

### ISSUE THREAD

**Issue by ranquild** [dart-lang/sdk#31253]
Immutable data are used heavily for web applications today, commonly with Elm-like (redux, ngrx, ...) architectures. Most common thing web developer is doing with data is creating a copy of it with some fields changed, usually propagated to the root of state tree. JavaScript has spread operator for this. There should be a easy way to use immutable data structures in Dart. I would like to have data classes (inspired by [Kotlin]( https://kotlinlang.org/docs/reference/data-classes.html)) in Dart. Possible API:
```
data class User {
  String name;
  int age;
}
```
Compiler assumes that all fields of data class are immutable. Compiler adds `equals` implementation based on shallow equals, `hashCode` implementation based on mix of all object fields, `toString` implementation of the form `<Type> <fieldN>=<valueN>`, and `copy` function to recreate object with some fields changed.

You may argue that there is already [Built value](https://github.com/google/built_value.dart) package that allows to do similar things, but we have many issues with package, mainly:
1. It requires writing a lot of boilerplate code
2. It requires running watcher/manual code generation during development.
3. It requires saving generated files to repository because code generation time is too large for big applications.

I have found that using built value actually decreases my productivity and I do my work faster with even manually writing builders for packages.

If data classes would be implemented on language level, it would increase developer productivity and optimizations can be made when compiling code to particular platform.

**Comment by eernstg** [comment]
No promises, but it's on the radar..

**Comment by freewind** [comment]
Or inline make it shorter?

`data class User(String name, int age)`

**Comment by dcovar** [comment]
Are there any updates on this enhancement? I'm currently working with Flutter, and having come from the Kotlin/Android world, this is something that would make the transition a lot nicer. Especially when creating ViewModels, or even simple data models, this would make it a lot easier.

**Comment by zoechi** [comment]
@dcovar don't expect anything short term. It won't be part of Dart 2. 
They might tackle it after Dart 2.

**Comment by fmatosqg** [comment]
Community could write a package similar to Lombok who autogenerates code from a valid annotated source code file.

https://projectlombok.org/

One more thing for the wishlist on https://github.com/flutter/flutter/issues/13607

**Comment by zoechi** [comment]
@fmatosqg https://pub.dartlang.org/packages/built_value

**Comment by Cat-sushi** [comment]
`User` should have implicit constructor `const User({this.name, this.age});`, correct?

**Comment by saolof** [comment]
One thing worth mentioning is that data classes can be viewed as an instance of a metaclass. If enough different kinds of special cased classes are proposed, at some point it might become better to add metaclass programming to the language, with a few individual cases of syntax sugar.

**Comment by andrewackerman** [comment]
I support this, but also suggest adding `toJson` and `fromJson` methods to the generated code so data class instances can be easily (de)serializable.

**Comment by i-schuetz** [comment]
@andrewackerman Data classes shouldn't have more than a generic minimum to be used as domain entities, this being `equals`/`hashCode`, `copy` and `toString`. Serialization isn't a universal requirement and even less to an industry standard (which not necessarily everyone wants to use / can become outdated) like JSON.

**Comment by eernstg** [comment]
For current activities in this direction, you may want to consider also issues like the following: dart-lang/language#117, dart-lang/language#125, dart-lang/language#225, dart-lang/language#308.

**Comment by andrewackerman** [comment]
@i-schuetz Then maybe there can be some optional attributes that can be added to the class declaration so these features can be added for people who need them? Serialization may not be a universal requirement but I'd bet that it would be needed often enough that people would want to at least have the option. Otherwise it would largely defeat the purpose of having a concise data class declaration syntax but then have to manually create the (de)serialization methods.

And it's not like it would need to serialize to actual JSON strings. It could serialize to `Map<String, dynamic>`, which is itself easily convertible to and from JSON.

**Comment by i-schuetz** [comment]
Maybe something generic along the lines of [Swift's `Codable`](https://developer.apple.com/documentation/swift/codable) could make sense, but this is an entirely different feature. Although `equals` and `toString` are convenience as well - I remember for example in Haskell this being solved via type classes (which to be implemented require only to write a word practically). I don't know which exact considerations Jetbrains did to shape data classes the way they did in Kotlin. It's probably something along the lines that `equals` and `toString` make sense _always_. Serialization, as you say, it's used only "often".

**Comment by dcov** [comment]
I agree with @i-schuetz that adding a `Codable` protocol is probably the best option. It could even make its way into Flutter plugins (and the framework itself), where the data you pass to the 'other side' has to be encoded first.

**Comment by kevmoo** [comment]
@leafpetersen @munificent @lrhn – should be moved to the language repo?

**Comment by leafpetersen** [comment]
> @leafpetersen @munificent @lrhn – should be moved to the language repo?

yes, thanks.
